### PR TITLE
feat(http): support of timeout for HTTP requests

### PR DESCRIPTION
**Pull request by templth**
Allow specifying a timeout to automatically abort requests through request
options.

Closes: https://github.com/angular/http/issues/74

### CODE REVIEW

File: modules/@angular/http/src/backends/jsonp_backend.ts [hunk 1]
```diff
@@ -108,20 +109,29 @@ export class JSONPConnection_ extends JSONPConnection {
...
+        this.cancel(script, onLoad, onError);
       };
     });
   }
 
+  cancel(script: any, onLoad, onError) {
+    this.readyState = ReadyState.Cancelled;
+    script.removeEventListener('load', onLoad);
+    script.removeEventListener('error', onError);
+    if (isPresent(script)) {
```
Comment (vicb) [inline modules/@angular/http/src/backends/jsonp_backend.ts]: the 2 previous lines look suspicious if `!isPresent(script)`

File: modules/@angular/http/src/backends/xhr_backend.ts [hunk 2]
```diff
@@ -37,6 +37,9 @@ export class XHRConnection implements Connection {
       if (isPresent(req.withCredentials)) {
         _xhr.withCredentials = req.withCredentials;
       }
+      if (isPresent(req.timeout)) {
+      _xhr.timeout = req.timeout;
```
Comment (vicb) [inline modules/@angular/http/src/backends/xhr_backend.ts]: missing WS

Comment (templth) [inline modules/@angular/http/src/backends/xhr_backend.ts]: Thanks for pointing this out!

File: modules/@angular/http/src/backends/jsonp_backend.ts (hunk 1 above)
Comment (templth) [inline modules/@angular/http/src/backends/jsonp_backend.ts]: Yes, you're right! It would be safer to move these two lines. Thanks!

Comment (templth) [inline modules/@angular/http/src/backends/jsonp_backend.ts]: Fixed

File: modules/@angular/http/src/backends/xhr_backend.ts (hunk 2 above)
Comment (templth) [inline modules/@angular/http/src/backends/xhr_backend.ts]: Fixed

File: modules/@angular/http/src/base_request_options.ts [hunk 3]
```diff
@@ -56,7 +56,11 @@ export class RequestOptions {
    * Enable use credentials for a {@link Request}.
    */
   withCredentials: boolean;
-  constructor({method, headers, body, url, search, withCredentials}: RequestOptionsArgs = {}) {
+  /**
+   * Timeout for a {@link Request}.
```
Comment (vicb) [inline modules/@angular/http/src/base_request_options.ts]: useless comment

File: modules/@angular/http/test/backends/xhr_backend_spec.ts [hunk 4]
```diff
@@ -602,6 +603,27 @@ export function main() {
...
+
+      it('should set timeout to 2000 when defined in request options',
+        inject([AsyncTestCompleter], (async: AsyncTestCompleter) => {
+          var statusCode = 200;
+          sampleRequest.timeout = 2000;
+          var mockXhr = new MockBrowserXHR();
+          var connection =
+              new XHRConnection(sampleRequest, mockXhr, new ResponseOptions({ status: statusCode }));
+          var responseHeaders = `X-Request-URL: http://somedomain.com
+           Foo: Bar`
```
Comment (vicb) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: is this needed

File: modules/@angular/http/test/backends/xhr_backend_spec.ts [hunk 5]
```diff
@@ -602,6 +603,27 @@ export function main() {
            existingXHRs[0].setStatusCode(statusCode);
            existingXHRs[0].dispatchEvent('load');
          }));
+
+      it('should set timeout to 2000 when defined in request options',
```
Comment (vicb) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: remove `to 2000`

File: modules/@angular/http/test/backends/xhr_backend_spec.ts [hunk 6]
```diff
@@ -602,6 +603,27 @@ export function main() {
...
+          var statusCode = 200;
+          sampleRequest.timeout = 2000;
+          var mockXhr = new MockBrowserXHR();
+          var connection =
+              new XHRConnection(sampleRequest, mockXhr, new ResponseOptions({ status: statusCode }));
+          var responseHeaders = `X-Request-URL: http://somedomain.com
+           Foo: Bar`
+
+          connection.response.subscribe((res: Response) => {
+            expect(res.url).toEqual('http://somedomain.com');
```
Comment (vicb) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: is this already tested elsewhere ?

File: modules/@angular/http/test/backends/xhr_backend_spec.ts [hunk 7]
```diff
@@ -602,6 +603,27 @@ export function main() {
...
+            expect(existingXHRs[0].timeout).toBeTruthy();
```
Comment (vicb) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: .toEqual ?

File: modules/@angular/http/src/base_request_options.ts (hunk 3 above)
Comment (templth) [inline modules/@angular/http/src/base_request_options.ts]: Do you mean the comment isn't necessary here? Or its content doesn't contain enough details? For the second, what about: "Number of milliseconds a request can take before automatically being terminated"?

Comment (vicb) [inline modules/@angular/http/src/base_request_options.ts]: I don't think the current comment bring any valuable info (nor `a request can take before automatically being terminated`). Maybe rename the property to `timeoutMs` ?

Comment (templth) [inline modules/@angular/http/src/base_request_options.ts]: Renaming to `timeoutMs` is fine for me! Do you expect me to rename the property everywhere? For the comment, I would propose "Timeout in milliseconds for the Request".

File: modules/@angular/http/test/backends/xhr_backend_spec.ts (hunk 5 above)
Comment (templth) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: Done

File: modules/@angular/http/test/backends/xhr_backend_spec.ts (hunk 4 above)
Comment (templth) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: Actually no

File: modules/@angular/http/test/backends/xhr_backend_spec.ts (hunk 7 above)
Comment (templth) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: Yes, `toEqual(2000)` is much better!

File: modules/@angular/http/test/backends/xhr_backend_spec.ts (hunk 6 above)
Comment (templth) [inline modules/@angular/http/test/backends/xhr_backend_spec.ts]: Yes, removed it.

File: modules/@angular/http/src/base_request_options.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/http/src/base_request_options.ts]: If you rename it to timeoutMs, the comment is useless the meaning is expressed by the name only

Comment (templth) [inline modules/@angular/http/src/base_request_options.ts]: I know that naming is something really important and I would like to be consistent with what is used for Angular 2. I saw that such property is called `timeout` in the Angular 1 [`$http`](https://docs.angularjs.org/api/ng/service/$http) object and it's the same for the XHR object.

Comment (vicb) [inline modules/@angular/http/src/base_request_options.ts]: call it timeout then, specify the unit in the comment.

Comment (templth) [inline modules/@angular/http/src/base_request_options.ts]: I updated the comments accordingly... Thanks!

**Comment by robwormald** [comment]
this can be solved downstream with composition, because Observables can be cancelled externally.

Something like

```js
import 'rxjs/add/operator/timeout'

http.get(url)
  .timeout(5000)
  .subscribe(...)
```
More idiomatic, imho.

**Comment by CaptainCodeman** [comment]
That wouldn't cancel the underlying http request though, just the subscription for it?

So the pending request could still be there and, I presume, the server wouldn't be able to check if the client was still listening and would send whatever it was going to (potentially something large?)

**Comment by robwormald** [comment]
@captaincodeman it would indeed cancel the underlying XHR, that's one of the primary reasons we opted for Observables in Http.

**Comment by CaptainCodeman** [comment]
It didn't look to me like the AjaxObservable from rxjs was being used (which appears to handle abort on unsubscribe).

I can't see how timeout causes the xhr abort to happen in the code before this PR.

**Comment by robwormald** [comment]
@CaptainCodeman .unsubscribe() is called by timeout internally, which calls abort() on the XHR - https://github.com/angular/angular/blob/master/modules/%40angular/http/src/backends/xhr_backend.ts#L102-L105

http://plnkr.co/edit/a5nqxlXFbZInyC7iKe4j?p=preview for an example

**Comment by templth** [comment]
@robwormald: I definitely agree with you. This also handle what we have in angular1: "timeout – {number|Promise} – timeout in milliseconds, or promise that should abort the request when resolved.". Should I close the PR?

**Comment by vicb** [comment]
ping @robwormald

**Comment by robwormald** [comment]
@templth yeah, i'm going to go ahead and close it. we'll add this sort of thing to docs. thanks again 👍

**Pull request closed without merging**